### ISSUE TITLE
Only one vmstorage node CPU suddenly becomes high in the cluster

### ISSUE THREAD

**Issue by petanne** [VictoriaMetrics/VictoriaMetrics#3425]
**Describe the bug**
Our cluster has 12 vmstorage (each 48 cores, 384GB memory and lots of free, 3% disk used), 4 vmselect, 4 vminsert, 4 vmagent, 1 vmalert. Only vmalert making read requests.
But one vmstorage node CPU suddenly becomes high without human operation, neither insert requests nor search requests increased. When we stop the vmalert, it recovered.
I have replaced the machine and double checked, it should not be a machine problem.

**To Reproduce**
Start vmalert again, nothing happened. But i have another way to reproduce:
Add a new vmalert and start it, one vmstorage CPU becomes high quickly, in about 1 minute.Then stop the problematic vmstorage, another vmstorage CPU becomes high quickly. When stop the second problematic vmstorage, another vmstorage will reproduce again.

**Expected behavior**
vmstorage CPU running normally.

**Logs**
no error or warning log in vmstorage.

**Screenshots**
stats
![20221201155703](https://user-images.githubusercontent.com/4476119/204997305-cb760d7c-93f0-4864-b27c-8ff8d89090e2.jpg)

vmstorage CPU
![20221201155115](https://user-images.githubusercontent.com/4476119/204996280-b504355b-98e3-4294-8aa3-8e29eed3c64d.jpg)

A test case, stop problematic vmstorage, looks like at least one vmstorage cpu must high?
![image](https://user-images.githubusercontent.com/4476119/205000348-1d4d6057-3b7c-4493-b8b3-aa4d70147dce.png)

I think may be something wrong in vmstorage? Because it's insert and search requests is balanced and unanimously.

rate(vm_vminsert_metrics_read_total[1m])
![20221201153535](https://user-images.githubusercontent.com/4476119/204993578-573acaef-888a-477c-8a17-c39d4e578bf3.jpg)

rate(vm_vmselect_rpc_requests_total[1m])
![20221201153925](https://user-images.githubusercontent.com/4476119/204994259-bafe0f22-34e8-479e-bafd-a3f41b686948.jpg)

vmstorage other metrics
![20221201155422](https://user-images.githubusercontent.com/4476119/204997101-51a288e1-41f8-4979-936d-feaf2522ba84.jpg)
![20221201155848](https://user-images.githubusercontent.com/4476119/204997689-0660693b-c5ed-49c9-8f37-62fefe703dbf.jpg)
![20221201155836](https://user-images.githubusercontent.com/4476119/204997695-cbc475ef-11b3-4be6-ba8c-97b489e56af9.jpg)
![20221201155824](https://user-images.githubusercontent.com/4476119/204997705-d78d7d2a-9658-4ebe-b788-03d3818f808d.jpg)
![20221201155811](https://user-images.githubusercontent.com/4476119/204997712-208c4a8d-7342-472f-9a4e-d3715dc4c82f.jpg)
![20221201155753](https://user-images.githubusercontent.com/4476119/204997716-88062931-0e62-45c4-81cb-a8496479f5ba.jpg)

problematic vmstorage go metric 
![20221201162108](https://user-images.githubusercontent.com/4476119/205002445-5903f64a-3ecf-4151-845c-d9bb0cfbd88e.jpg)
![20221201162218](https://user-images.githubusercontent.com/4476119/205002452-eb4a7137-63c6-41f1-bd32-4d0e6985c6f1.jpg)
![20221201162142](https://user-images.githubusercontent.com/4476119/205002457-dad45134-7887-462a-afd8-5cc707929764.jpg)

pprof and trace when CPU high
[pprof_and_trace.zip](https://github.com/VictoriaMetrics/VictoriaMetrics/files/10130400/pprof_and_trace.zip)

**Version**
```
vmstorage-20221014-073231-tags-v1.82.1-cluster-0-gb942dd77c
```

**Used command-line flags**
```
/app/vmstorage/vmstorage-prod -memory.allowedPercent=60 -search.maxUniqueTimeseries=30000000 -dedup.minScrapeInterval=30s -retentionPeriod=2y -storage.minFreeDiskSpaceBytes=10GB -storageDataPath=/data/vmstorage-data
```

**Comment by hagen1778** [comment]
Hi @petanne! Thanks for detailed report!

> Our cluster has 12 vmstorage (each 48 cores, 384GB memory and lots of free, 3% disk used)

How many vmstorage nodes are on the same instance? Are there other heavy processes running on the same instance?
Could you please show RSS memory usage on problematic storage node and its memory limit?
```
max_over_time(process_resident_memory_bytes{}[10m])

vm_available_memory_bytes{}
```

**Comment by petanne** [comment]
Hi @hagen1778! Thanks for reply!
One vmstorage process on each physical machine. There are no any other heavy process. 

Red is problematic time range.
![image](https://user-images.githubusercontent.com/4476119/205259274-3ee41766-9821-4530-8c52-0fbb4cf1a8ef.png)

Some metrics from node_exporter.
![image](https://user-images.githubusercontent.com/4476119/205260537-7ce51084-8a3f-4003-982e-f5901ebc4a4d.png)
![image](https://user-images.githubusercontent.com/4476119/205260671-8d59ed99-52df-43d0-ba5f-bfe0bb5e4ef1.png)

I have 10 nodes prometheus, different targets and alert rules on each prometheus, they run for more than a year. In VictoriaMetrics cluster, both targets and alert rules are same as prometheus.

Does some special metrics or query can cause this problem?

**Comment by hagen1778** [comment]
> Does some special metrics or query can cause this problem?

It could. But I'd expect it to happen only when vmstorage experiences memory shortage. Usually, such extensive and long CPU increase is a sign of [page cache thrashing](https://www.techtarget.com/searchsoftwarequality/definition/cache-thrash), which happens due to memory shortage or by other process on the instance doing write/read heavy operations and polluting the cache. 
But according to your screenshots, vmstorage uses like 10% of available memory so this shouldn't be the case.

 Do you have the following panels on your dashboard?
<img width="1283" alt="image" src="https://user-images.githubusercontent.com/2902918/205268740-40246120-fd18-4047-a1d9-d9c0b006120b.png">
They might have a bit different names, but still can be used to show how heavy were queries at that time. If you don't have these panels, please find them [here](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/dashboards/victoriametrics-cluster.json).

**Comment by petanne** [comment]
> Do you have the following panels on your dashboard?

Red is problematic time range.
![image](https://user-images.githubusercontent.com/4476119/205279926-46425fce-88dd-45cc-b7e3-da2021615d9a.png)

Another test case, i try increase `-search.maxConcurrentRequests` for vmselect, problem still occurs.
![image](https://user-images.githubusercontent.com/4476119/205283596-bc9a5f4d-18dd-44c3-9ef6-63a086a04f29.png)

I found in pprof that the `lib/storage.(*TSID).Less` method is called more times than normal. What would cause this?

**Comment by petanne** [comment]
We try set `-search.maxUniqueTimeseries` for vmstorage to 100000, problem still occurs.

**Comment by hagen1778** [comment]
> I found in pprof that the lib/storage.(*TSID).Less method is called more times than normal. What would cause this? 

More times or just takes more time?
This method supposed to be in the top of CPU usage for all vmstorage nodes, since it is responsible for filtering data parts during lookups. The more read pressure is there the higher is impact on parts filtering. 

Could you please run the following query for problematic storage?
```
max(
    vm_cache_size_bytes{job=~"$job_storage", instance=~"$instance"} 
    /
    vm_cache_size_max_bytes{job=~"$job", instance=~"$instance"}
) by(type)
```

It supposed to show cache usage % by type for the vmstorage node.

**Comment by petanne** [comment]
> More times or just takes more time?

In problematic time range,  `lib/storage.(*TSID).Less` is 30% flat in the problematic vmstorage, but other vmstorage is 5%.

This is cache usage % by type for the vmstorage node, 13.vmstorage is problematic node, 1.vmstorage CPU is normal. All of other storage are like 1.vmstorage.
![image](https://user-images.githubusercontent.com/4476119/205427290-a3c38889-8556-4456-88c7-72da9565f4b4.png)

The cache which type is `indexdb/tagFilters`, 99% usage on each vmstorage node. Is this normal?

**Comment by valyala** [comment]
The commit 8e9822bc7f0f0591555be4faa76dd5af431e2000 optimizes the hostspot at `TSID.Less()` function.
@petanne , could you try building `vmstorage` from this commit according to [these instructions](https://docs.victoriametrics.com/Cluster-VictoriaMetrics.html#building-from-sources) and verifying whether the `TSID.Less()` call is removed from CPU profile collected from the updated `vmstorage`? It is safe running `vmstorage` from this commit in the cluster with other `v1.82.1` components.

**Comment by petanne** [comment]
@valyala, thanks! It worked! 

The `TSID.Less()` flat in the CPU pprof is 3.87% now, it was 5% when normal before.
![image](https://user-images.githubusercontent.com/4476119/205626003-81bd18c2-6c7b-4137-935c-f37720505f74.png)

Expecting for release.
This time the sudden problem occurred in 2022-12-01 00:05 UTC, when the partition for this month is created. Could you tell me why and will it show up next month?

Victoria is great! Where can I learn `partition` and `block`? What happens during a search request in vmstorage?

**Comment by petanne** [comment]
@valyala @hagen1778 
Bad news, problem happened again, and other problems arised. When we do stress tests and outage drills.

### Version
vmstorage: [8e9822b](https://github.com/VictoriaMetrics/VictoriaMetrics/commit/8e9822bc7f0f0591555be4faa76dd5af431e2000)
vmagent: [v1.84.0-cluster](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.84.0-cluster)
vminsert: [v1.84.0-cluster](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.84.0-cluster)
vmselect: [v1.84.0-cluster](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.84.0-cluster)
vmalert: [v1.84.0-cluster](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.84.0-cluster)

### Timeline
Yesterday,  Set `-search.disableCache=true` in vmselect, and change `-evaluationInterval=30s` to `10s` in vmalert. it works fine until the next operation.

17:45 Reboot 13.vmstorage server. The 13.vmstorage has taken up 90% of the CPU usage since the process started, until we stoped all of 2 vmalert processes when 20 minutes after. __This is question 1.__

18:30 Start 2 vmalert processes. Everything is ok.

19:18 Reboot 1.vmstorage server. The 13.vmstorage CPU usage increase 90% immediately. __This is question 2.__
19:21 1.vmstorage process is auto started, but its CPU usage increase 90% immediately, and 13.vmstorage CPU usage returns to normal. __This is question 3.__
19:18 ~ 19:26, vmagent and vminsert network traffic is gradually drop to zero, then gradually recover, more serious. __This is question 4.__

19:32 Stop all vmalert processes. 1.vmstorage CPU usage is returns to normal. And vmagent and vminsert network traffic increased by 2 times. Looks like it's blocked by 1.vmstorage. __This is question 5.__

19:42 Change `-evaluationInterval=10s` to `30s` then start 2 vmalert processes. The 13.vmstorage CPU usage is fluctuates frequently between 15% and 30%, other vmstorage nodes behave normally until 22:35. __This is question 6.__

22:35 Nobody does anything, 13.vmstorage CPU burst to 90%, as said at the beginning of this issue. __This is question 7.__

### Dashboard
Red is 13.vmstorage. Green is 1.vmstorage. Blue is 2.vmstorage, normal example for comparison.
![image](https://user-images.githubusercontent.com/4476119/206670528-7ec39e31-e1e3-4695-9da0-6f06e1db494e.png)

Network traffic for one vmagent node.
![image](https://user-images.githubusercontent.com/4476119/206671341-1046623e-f5c7-48e7-8987-3356c58ae761.png)

***
I read the official documentation and part of the source code, but still can't answer the above 7 questions. Maybe I'm about to lose victoria.

**Comment by hagen1778** [comment]
> This is question 1.

Does pprof shows the same issue?

Just for context, we discussed the issue with @valyala before he submitted a fix. The fix isn't fixing anything, it is a workaround for an issue we suspect happens in your setup. If you take a look at the profile you provided initially, you'd find that majority of time is spent on a simple comparison operation:
```
         .          .    104:func (t *TSID) Less(b *TSID) bool {
         .          .    105:   // Do not compare MetricIDs here as fast path for determining identical TSIDs,
         .          .    106:   // since identical TSIDs aren't passed here in hot paths.
   393.32s    399.16s    107:   if t.AccountID != b.AccountID {
         .          .    108:           return t.AccountID < b.AccountID
         .          .    109:   }
 
```
While this operation is supposed to be very lightweight, it takes too much time in your case. The assumption is that `b *TSID` doesn't fit the CPU cache, and the process takes extra time to extract it from memory. We don't know why it happens yet, it is a pretty rare situation across all the installations we had a chance to investigate. 
So far, we think that adding more storage shards could help to stabilize the situation. You can significantly reduce the allocated resource per-node in favor of using more instances with lower resources for vmstorage layer.

**Comment by petanne** [comment]
> Does pprof shows the same issue?

Sorry for not keeping the scene and save pprof. This problem is easy to reproduce and the phenomenon is consistent, So I think the reason is the same too.

Based on our capacity assessment and machine resources, we use 12 vmstorage server,  each 48 cores, 384GB memory and 15TB disk. Although the machine configuration is high, but the data volume and usage rate are low now. I'm sorry there are no more machines to replace at this time.

I am not using multi-tenancy, each `AccountID` is default value or 0. Could this cause some unexpected behavior?

**Comment by petanne** [comment]
### New test case
Only start one vmalert with `-evaluationInterval=10s`, 13.vmstorage CPU is 90%. `storage.(*TSID).Less` called from `storage.(*partSearch).nextBHS` is biggest CPU usage again.

[13.error.cpu.pprof.12.9.zip](https://github.com/VictoriaMetrics/VictoriaMetrics/files/10194280/13.error.cpu.pprof.12.9.zip)
```
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*partSearch).nextBHS
/root/VictoriaMetrics/lib/storage/tsid.go
  Total:     177.22s    178.12s (flat, cum) 14.76%
    107      169.88s    170.75s           if t.AccountID != b.AccountID {
```

Does nextBHS need optimization for `Less`?
```
func (ps *partSearch) nextBHS() bool {
	for len(ps.metaindex) > 0 {
		// Optimization: skip tsid values smaller than the minimum value
		// from ps.metaindex.
		for ps.BlockRef.bh.TSID.Less(&ps.metaindex[0].TSID) {
			if !ps.nextTSID() {
				return false
			}
		}
```

As a supplementary note, in the above case, the replication of series data on 13.vmstorage will be on the 1.vmstorage.

I want to known why are so many data blocks retrieved on 13.vmstorage? I checked all alert queries and found no suspicious.

**Comment by hagen1778** [comment]
> Based on our capacity assessment and machine resources, we use 12 vmstorage server, each 48 cores, 384GB memory and 15TB disk.

What I'm saying is that you can replace 48 cores machine with 3 machines 16 cores each and if the assumption with CPU cache is correct - the problem will be solved. But I understand it could be complicated when not running in cloud.

> Could this cause some unexpected behavior?

No.

> I want to known why are so many data blocks retrieved on 13.vmstorage?  

It is unclear. We haven't observed such situations before. Even for a much bigger load and less hardware. Could you please share an example of the alert queries you run?

**Comment by petanne** [comment]
Yes, we use physical machines, not cloud. Sorry, there are not enough machines to replace 48 cores with 3 * 16 cores.

> Could you please share an example of the alert queries you run?

Here are some typical alert queries. They have been working in our Prometheus for many years.
```
up{port="9100"} == 0
```

```
sum(increase(log_monitor_counter{level="err"}[1m])) >= 10 
or 
log_monitor_counter{level="err"} >= 10 
    unless log_monitor_counter{level="err"} offset 1m 
    unless log_monitor_counter{level="err"} offset 10m
```

```
sum(rate(nginx_http_upstream_request_times_count{endpoint=~".*xxxx"}[1m])) / sum(rate(nginx_http_upstream_request_times_count{endpoint=~".*xxxx"}[1m] offset 1d)) < 0.8 
and 
sum(rate(nginx_http_upstream_request_times_count{endpoint=~".*xxxx"}[1m])) / sum(rate(nginx_http_upstream_request_times_count{endpoint=~".*xxxx"}[1m] offset 1w)) < 0.8
```

```
rate(is_machine_alive_change{}[3m]) < 0.002 and (hour() > 19  or hour() < 16)
```

**Comment by hagen1778** [comment]
> Set -search.disableCache=true in vmselect

Why would you do that? Disabling caches also increases the amount of work data blocks scanning. Normally, some previously read blocks are cached in memory to reduce the lookups.

> 17:45 Reboot 13.vmstorage server. The 13.vmstorage has taken up 90% of the CPU usage since the process started, until we stoped all of 2 vmalert processes when 20 minutes after. This is question 1.

It could be that after the restart caches for 13.vmstorage didn't fill up caches enough to remain efficient. It usually takes not that much time, but here we assume that load on the node is too high (due to unknown reason so far).

> 19:18 Reboot 1.vmstorage server. The 13.vmstorage CPU usage increase 90% immediately. This is question 2.

When one storage goes off, vminserts start to re-route data from the "dead" storage to other "healthy" storages. Re-routing always increases the load on the rest of the nodes. And if our theory of vmstorage-running-on-the-edge is correct, then this additional load only increases the probability of entering this state.

> 19:21 1.vmstorage process is auto started, but its CPU usage increase 90% immediately, and 13.vmstorage CPU usage returns to normal. This is question 3.

It is likely the same situation as in q1 - restoring the caches under read queries load. 13.vmstorage stopped receiving additional load from rerouting and got back to normal.

> 19:18 ~ 19:26, vmagent and vminsert network traffic is gradually drop to zero, then gradually recover, more serious. This is question 4.

There are two types of re-routing in vm cluster:
1. when vmstorage becomes unreachable (goes off or network split)
2. when vmstorage becomes slow (performance degradation) - this type of rerouting is disabled by default. We recommend enabling it if cluster contains > 10 storage nodes.

In your case, re-rerouting p2 was disabled. So when one vmstorage became slow - the ingestion speed was limited by the slowest vmstorage node in the cluster. You need to enable re-routing by setting `-disableRerouting=false` on vminsert nodes.

> 19:32 Stop all vmalert processes. 1.vmstorage CPU usage is returns to normal. And vmagent and vminsert network traffic increased by 2 times. Looks like it's blocked by 1.vmstorage. This is question 5.

Read load was reduced, so 1.vmstorage got back to normal for processing incoming data.


> 19:42 Change -evaluationInterval=10s to 30s then start 2 vmalert processes. The 13.vmstorage CPU usage is fluctuates frequently between 15% and 30%, other vmstorage nodes behave normally until 22:35. This is question 6.
22:35 Nobody does anything, 13.vmstorage CPU burst to 90%, as said at the beginning of this issue. This is question 7.

Back to my theory, I assume vmstorage nodes in this setup for some reason run-on-the-edge in a matter of compute resources. We have seen similar behaviour but only for setups where memory, cpu or disk were very close to 90-99% of utilization. Your case is different. According to the provided data, there are plenty of non-utilized resources. So the only assumption so far is the CPU cache issues. When read load crosses the "edge" - vmstorage falls into vicious circle where normal operations start to take too much time. This is when it enters into 90% CPU utilization.

Let's try to verify this theory by checking the CPU cache miss ratio in normal operation and faulty operation. Go processes do not have access to this info, so I'd like to ask you to take the measurements on your own. Please, take a look at this [discussion](https://stackoverflow.com/questions/18172353/how-to-catch-the-l3-cache-hits-and-misses-by-perf-tool-in-linux). This [comment](https://stackoverflow.com/a/46439957/1645633) provides an example of how to measure the misses in L3 cache - would be great if you can make the measurements before and during the incident. Please note the `sleep` command in the end, it is likely an interval of the measurement. It would be better to extend it during measurements.

> Yes, we use physical machines, not cloud. Sorry, there are not enough machines to replace 48 cores with 3 * 16 cores.

Understandable. If that's an option, you can try using the cloud for running the test once again, but on the cloud's hardware. You can rule out the hardware issue or try different topologies by using cloud, even if temporary. 

As I said earlier, we haven't observed the issue like yours before. So if you're using some kind of benchmark for generating the ingested data - let us know, so we can try to reproduce it.

> Here are some typical alert queries. They have been working in our Prometheus for many years.

The queries look normal. Thank you!

**Comment by petanne** [comment]
Thank you for your careful answer.

> Why would you do that? Disabling caches also increases the amount of work data blocks scanning. Normally, some previously read blocks are cached in memory to reduce the lookups.

In order to do a pressure test for vmstorage. The cache is enabled before this test.

### Command for system-wide L3 cache miss rate
```
$ perf stat -a -e LLC-loads,LLC-load-misses,LLC-stores,LLC-store-misses,LLC-prefetch-misses sleep 10
```

On 13.vmstorage, when CPU is 90%.
```
 Performance counter stats for 'system wide':

     1,672,017,968      LLC-loads                                                     (50.02%)
       566,498,839      LLC-load-misses           #   33.88% of all LL-cache hits     (50.01%)
       307,252,131      LLC-stores                                                    (50.01%)
       130,520,462      LLC-store-misses                                              (50.01%)
   <not supported>      LLC-prefetch-misses

      10.009098222 seconds time elapsed
```

On 2.vmstorage, at the same time.
```
 Performance counter stats for 'system wide':

     1,289,752,696      LLC-loads                                                     (50.00%)
       378,444,294      LLC-load-misses           #   29.34% of all LL-cache hits     (50.00%)
       323,470,831      LLC-stores                                                    (50.00%)
       107,277,991      LLC-store-misses                                              (50.00%)
   <not supported>      LLC-prefetch-misses

      10.001611238 seconds time elapsed
```

As a comparison, the following is when only one vmalert process is started and the CPU is normal.
On 13.vmstorage.
```
 Performance counter stats for 'system wide':

     1,149,413,385      LLC-loads                                                     (50.00%)
       336,088,891      LLC-load-misses           #   29.24% of all LL-cache hits     (50.00%)
       310,156,805      LLC-stores                                                    (50.00%)
       100,309,852      LLC-store-misses                                              (50.00%)
   <not supported>      LLC-prefetch-misses

      10.001550442 seconds time elapsed
```

On 2.vmstorage.
```
 Performance counter stats for 'system wide':

     1,398,281,187      LLC-loads                                                     (50.00%)
       341,360,772      LLC-load-misses           #   24.41% of all LL-cache hits     (50.00%)
       295,972,084      LLC-stores                                                    (50.00%)
        93,261,754      LLC-store-misses                                              (50.00%)
   <not supported>      LLC-prefetch-misses

      10.001704804 seconds time elapsed
```

How can I see how many data blocks were scanned by a query? I can't find it in the [query-tracing](https://docs.victoriametrics.com/#query-tracing).

**Comment by petanne** [comment]
From this issue until now, we use `vmctl` to import Prometheus historical data, and limit the import rate. There is only a short time in between when no import is performed.

### Import command
```
/app/vmctl/vmctl-prod 
    prometheus 
    --vm-addr=http://vminsert_addr/ 
    --prom-snapshot=/data/transit/p14/prometheus/snapshots/20221206T035732Z-6aeb8e98f4c7f371/ 
    --vm-concurrency=48 
    --vm-batch-size=20480000 
    --prom-concurrency=48 
    --prom-filter-time-end=2022-11-08T00:07:01Z 
    --vm-account-id=0 
    --vm-rate-limit=30720000 
    -s=true
```
This command is executed on a separate new machine. I don't know if this will make a difference.

**Comment by petanne** [comment]
> So if you're using some kind of benchmark for generating the ingested data - let us know, so we can try to reproduce it.

Sorry, no benchmark data.The time series data is scraped by vmagent from our production environment.

**Comment by petanne** [comment]
Hi, @valyala , Could you help analyze it?

**Comment by valyala** [comment]
@petanne , could you try building `vmstorage` from the commit fbeebe4869892a9dd3c684da5fbaa0a6c511a651 and verify whether this helps resolving the issue? See [build instructions](https://docs.victoriametrics.com/Cluster-VictoriaMetrics.html#building-from-sources). It is safe running `vmstorage` from this commit with other cluster components built from older commits (e.g. previous releases).

**Comment by valyala** [comment]
> Where can I learn partition and block? What happens during a search request in vmstorage?

Please take a look at https://docs.victoriametrics.com/#storage and https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3268#issuecomment-1291249751

**Comment by petanne** [comment]
> @petanne , could you try building `vmstorage` from the commit [fbeebe4](https://github.com/VictoriaMetrics/VictoriaMetrics/commit/fbeebe4869892a9dd3c684da5fbaa0a6c511a651) and verify whether this helps resolving the issue? 

I build [fbeebe4](https://github.com/VictoriaMetrics/VictoriaMetrics/commit/fbeebe4869892a9dd3c684da5fbaa0a6c511a651), the problem still occurs.
![image](https://user-images.githubusercontent.com/4476119/207796029-a0fccb3f-e790-4f30-b04f-6c3fed890013.png)

I found `blocks_read` on 13.vmstorage is high than other vmstorage, when 13.vmstorage CPU is high. I don't know if this has a clue.
```
rate(vm_vmselect_rpc_requests_total{srv="$srv", action="search"}[1m])
sum(rate(vm_vmselect_metric_blocks_read_total[1m])) by (instance) / sum(rate(vm_vmselect_rpc_requests_total{action="search"}[1m])) by (instance)
```
![image](https://user-images.githubusercontent.com/4476119/207796562-3c6c1e74-fdbe-49e6-9d8e-d77da609ed46.png)

![image](https://user-images.githubusercontent.com/4476119/207797344-ff90c625-28c8-4ffa-a597-cac42207dad1.png)

**Comment by valyala** [comment]
@petanne , could you share graphs for the following queries on the same time range as graphs above if these graphs expose some anomalies around the CPU usage spikes time frames for the instance `13.vmstorage.monitor.oss`, while the remaining instances have no any anomalies?

- `vm_parts{type="storage/inmemory"}`
- `vm_parts{type="storage/small"}`
- `vm_parts{type="indexdb/inmemory"}`
- `vm_parts{type="indexdb/file"}`
- `rate(process_io_storage_read_bytes_total{instance=~".+vmstorage.+"})`
- `rate(process_io_storage_written_bytes_total{instance=~".+vmstorage.+"})`

**Comment by petanne** [comment]
I reproduced this problem again.

1. Stop all of 2 vmalert instance.
2. Upgrade all vmstorage to https://github.com/VictoriaMetrics/VictoriaMetrics/commit/fbeebe4869892a9dd3c684da5fbaa0a6c511a651 .
3. Start one vmalert instance. Red in graphs.

![image](https://user-images.githubusercontent.com/4476119/208075650-475bb281-ff93-4864-83bc-abef11b60b79.png)
![image](https://user-images.githubusercontent.com/4476119/208075787-bffa411a-d343-46cf-ad21-a4a71c6dc8bf.png)
![image](https://user-images.githubusercontent.com/4476119/208076280-82ba0ad2-0086-4bd3-842b-0406fc82353a.png)

**Comment by valyala** [comment]
@petanne , thanks for the provided graphs! Could you try building and running `vmstorage` from the commit 8c08d625eeb4c5b224b031f27693b62bc4403602 ?

**Comment by valyala** [comment]
@petanne , the commit 8c08d62 has been included in [v1.85.2](https://github.com/VictoriaMetrics/VictoriaMetrics/releases), so you can try the official `vmstorage` from the `v1.85.2-cluster` tag instead of building it on your own. It also included an important bugfix for https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3502

**Comment by petanne** [comment]
Add to last comment, stopped the data import and the CPU problem still occurs.

**Comment by valyala** [comment]
@petanne , thanks for providing useful graphs! It looks like the CPU usage highly correlates the the number of parts of the `storage/inmemory` type. Big number of data parts may significantly increase CPU usage during queries. The commit 8dc04a86f6be9be46796e9980944a7ced793a05d should reduce the maximum number of data parts under high load by increasing the chances of parts' merge in foreground during data ingestion (aka assisted merge).

@petanne , could you build `vmstorage` from the commit 8dc04a86f6be9be46796e9980944a7ced793a05d and verify whether this helps reducing CPU usage under your workload?

**Comment by ikenchina** [comment]
Hi
After We increased size of shards of cache, the problem was resovled.
So, we suspect that cache size led CPU usage to 100%, there wasn't enough CPU to merge inmemory parts.
But, we still don't know what caused increase in the size of cache?

**Comment by ikenchina** [comment]
@valyala 
I was able to reproduce the problem when i ran a stress testing at 3x the traffic(v1.85.2).
But i built a binary as your commit, the problem was worse, it was reproduced at 1x the traffic.

**Comment by valyala** [comment]
@petanne , could you try building VictoriaMetrics cluster components from the commit 103dfd0525fc4c78b6ad55a37e9be2aa45aa840f ? It contains additional performance fixes on top of those mentioned at https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3425#issuecomment-1366967143 .

**Comment by ikenchina** [comment]
Problem is fixed after i commented the WaitIfNeeded function in PaceLimiter.

**Comment by valyala** [comment]
> Problem is fixed after i commented the WaitIfNeeded function in PaceLimiter.

Thanks! Then the issue must be fixed in the commit 09d7fa27372fc9252e01de00a863f3b7e075b4a5 , which removes the PaceLimiter . This commit is included in [the latest release](https://docs.victoriametrics.com/CHANGELOG.html).

**Comment by valyala** [comment]
FYI, one `vmstorage` node could suddenly use more CPU than other `vmstorage` nodes when it couldn't keep up with background merging of created `indexdb` parts as described at https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5190 . This issue has been addressed in [v1.97.0](https://docs.victoriametrics.com/CHANGELOG.html). Closing the issue as fixed then.

@petanne , please create a new issue and post a link to this one there if you still see sudden CPU usage spikes at some `vmstorage` nodes in VictoriaMetrics cluster running [v1.97.0](https://docs.victoriametrics.com/CHANGELOG.html) or newer release.
